Update no_trainer.py scripts to include accelerate gradient accumulation wrapper

Rasmusafj opened the pull request:
# What does this PR do?
Updates no_trainer.py scripts to use the new gradient accumulation wrapper feature from accelerate according to #18436. 

<!--
Congratulations! You've made it this far! You're not quite done yet though.

Once merged, your PR is going to appear in the release notes with the title you set, so make sure it's a great title that fully reflects the extent of your awesome contribution.

Then, please replace this with a description of the change and which issue is fixed (if applicable). Please also include relevant motivation and context. List any dependencies (if any) that are required for this change.

Once you're done, someone will review your PR shortly (see the section "Who can review?" below to tag some potential reviewers). They may suggest changes to make the code even better. If no one reviewed your PR after a week has passed, don't hesitate to post a new comment @-mentioning the same persons---sometimes notifications get lost.
-->

<!-- Remove if not applicable -->

## Before submitting
- [ N] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ Y] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ N] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [N ] Did you write any new necessary tests?


## Who can review?
@sgugger

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

<!-- Your PR will be replied to more quickly if you can figure out the right person to tag with @

 If you know how to use git blame, that is the easiest way, otherwise, here is a rough guide of **who to tag**.
 Please tag fewer than 3 people.

Models:

- albert, bert, xlm: @LysandreJik
- blenderbot, bart, marian, pegasus, encoderdecoder,  t5: @patrickvonplaten, @patil-suraj
- longformer, reformer, transfoxl, xlnet: @patrickvonplaten
- fsmt: @stas00
- funnel: @sgugger
- gpt2: @patrickvonplaten, @LysandreJik
- rag: @patrickvonplaten, @lhoestq
- tensorflow: @LysandreJik

Library:

- benchmarks: @patrickvonplaten
- deepspeed: @stas00
- ray/raytune: @richardliaw, @amogkam
- text generation: @patrickvonplaten
- tokenizers: @n1t0, @LysandreJik
- trainer: @sgugger
- pipelines: @LysandreJik

Documentation: @sgugger

HF projects:

- datasets: [different repo](https://github.com/huggingface/datasets)
- rust tokenizers: [different repo](https://github.com/huggingface/tokenizers)

Examples:

- maintained examples (not research project or legacy): @sgugger, @patil-suraj
- research_projects/bert-loses-patience: @JetRunner
- research_projects/distillation: @VictorSanh

 -->

Rasmusafj committed:
Added accelerate gradient accumulation wrapper to run_image_classification_no_trainer.py example script

Rasmusafj committed:
make fixup changes

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_18473). All of your documentation changes will be reflected on that endpoint.

sgugger commented on examples/pytorch/image-classification/run_image_classification_no_trainer.py:
```diff
@@ -213,7 +213,13 @@ def main():
     # in the environment
     accelerator = (
-        Accelerator(log_with=args.report_to, logging_dir=args.output_dir) if args.with_tracking else Accelerator()
+        Accelerator(
+            log_with=args.report_to,
+            logging_dir=args.output_dir,
+            gradient_accumulation_steps=args.gradient_accumulation_steps,
+        )
+        if args.with_tracking
+        else Accelerator(gradient_accumulation_steps=args.gradient_accumulation_steps)
```
Maybe define `log_kwargs` in an if/else then do just one call to `Accelerator`?

Rasmusafj replied:
I tried doing what you describe. Let me know if this is correct. :)

sgugger commented on examples/pytorch/image-classification/run_image_classification_no_trainer.py:
```diff
@@ -385,7 +391,7 @@ def collate_fn(examples):
         optimizer=optimizer,
         num_warmup_steps=args.num_warmup_steps,
-        num_training_steps=args.max_train_steps,
+        num_training_steps=math.ceil(len(train_dataloader)) * args.num_train_epochs,
```
```suggestion
        num_training_steps=len(train_dataloader) * args.num_train_epochs,
```
This length should be an integer.

Also we are ignoring the `max_steps` argument set by the user here, I don't think this change should be made.

Rasmusafj replied:
Yea, I see. 

So, now i've added that its `len(train_dataloader) * args.num_train_epochs` and `args.max_train_steps * args.gradient_accumulation_steps` in case user provided max_train_steps.

sgugger commented on examples/pytorch/image-classification/run_image_classification_no_trainer.py:
```diff
@@ -467,17 +473,20 @@ def collate_fn(examples):
                 lr_scheduler.step()
                 optimizer.zero_grad()
+
+            if accelerator.sync_gradients:
```
I think it's going to be easier to just always update, at least for now.

Rasmusafj replied:
If we always update and we have gradient_accumulation_steps > 1, the progress bar will act wrong i.e. increment fast and then stop while still performing updates.

This is because if max_train_steps is not given as an argument, its calculated based on dividing with gradient_accumulation_steps.

muellerzr replied:
Excellent point. I think the proper solution here would be a small wrapper around the progress bar (on top of the new one Accelerate just added) that is tied to the `GradientState` similar to how `optimizer.step()` works, where if the gradients should be synchronized then we assume a step has been performed and the progress bar moves. Otherwise it does nothing. 

But for now what you have here does look right to me.

sgugger reviewed (commented):
I think we need also change the definition of `num_update_steps_per_epoch` at line 385 and also multiply `max_steps` by the number of gradient accumulation steps, as discussed on the issue.

Thanks a lot for your PR!

Rasmusafj committed:
PR comments

Rasmusafj committed:
changed input to Acceletor based on PR comment, ran make fixup

sgugger reviewed (approved):
LGTM, let me know if you want to do more examples in the same PR or if this is good to merge!

Rasmusafj committed:
Added comment explaining the sync_gradients statement

Rasmusafj committed:
Fixed lr scheduler max steps

Rasmusafj committed:
Changed run_clm_no_trainer.py script to use accelerate gradient accum wrapper

Rasmusafj committed:
Fixed all scripts except wav2vec2 pretraining to use accelerate gradient accum wrapper

sgugger commented:
Let us know when the PR is ready for review (it's in draft mode right now) so we can go ahead and merge.

Rasmusafj committed:
Added accelerate gradient accum wrapper for wav2vec2_pretraining_no_trainer.py script

Rasmusafj committed:
make fixup and lr_scheduler step inserted back into run_qa_beam_search_no_trainer.py

Rasmusafj commented:
@sgugger  @muellerzr I changed PR to review. :)

sgugger commented on examples/pytorch/question-answering/run_qa_beam_search_no_trainer.py:
```diff
@@ -739,8 +747,8 @@ def create_and_fill_np_array(start_or_end_logits, dataset, max_len):
         optimizer=optimizer,
-        num_warmup_steps=args.num_warmup_steps,
-        num_training_steps=args.max_train_steps,
+        num_warmup_steps=args.num_warmup_steps * num_update_steps_per_epoch,
+        num_training_steps=args.max_train_steps * num_update_steps_per_epoch,
```
This should be `args.gradient_accumulation_steps`  here.

Rasmusafj replied:
fixed. :)

sgugger commented on examples/pytorch/speech-pretraining/run_wav2vec2_pretraining_no_trainer.py:
```diff
@@ -368,7 +368,7 @@ def main():
 
     # Initialize the accelerator. We will let the accelerator handle device placement for us in this example.
-    accelerator = Accelerator()
+    accelerator = Accelerator(gradient_accumulation_steps=args.gradient_accumulation_steps)
```
This one does a lot more things in its training loop, so would leave it as it was as I'm more afraid to break anything.

Rasmusafj replied:
Removed changes

sgugger reviewed (commented):
Thanks a lot, I wouldn't touch the Wav2Vec2 pretraining script as it is very different from the others and I'd need the speech team to validate those changes (you can make another PR with just it, but I think it's also fine to just leave it as is).

Apart from a typo, all the other ones look good to me!

Rasmusafj committed:
removed changes to run_wav2vec2_pretraining_no_trainer.py script and fixed using wrong constant in qa_beam_search_no_trainer.py script

Rasmusafj commented:
@sgugger I removed changes to wav2vec script and fixed the using of wrong constant. 

Feel sure to merge if its fine. Its going to be squash merged right? or do i need to rebase and squash before merge?

sgugger commented:
We squash indeed. Thanks again for your contribution!
